File: slides/04-multivariate-first-order/rsrc/negative_gradients.py
import numpy as np
import matplotlib.pyplot as plt
import os
dirpath = os.path.dirname(__file__)

# from: Mathematics for Machine Learning, p227, Figure 7.2

x = np.linspace(-6, 2, 50)
y = x**4 + 7*(x**3) + 5*(x**2) - 17*x + 3

# function
nsx = -4.48
nsy = nsx**4 + 7*(nsx**3) + 5*(nsx**2) - 17*nsx + 3

# global minimum
ns = [nsx, nsy]
point1 = [nsx, -60]
point1x = [ns[0], point1[0]]
point1y = [ns[1], point1[1]]
point2 = [-6.5, nsy]
point2x = [ns[0], point2[0]]
point2y = [ns[1], point2[1]]



plt.figure()
plt.plot(x, y, color='red', label='x$^4$ + 7x$^3$ + 5x$^2$ - 17x +3')
plt.plot(point1x, point1y, linestyle='--', color='blue')
plt.plot(point2x, point2y, linestyle='--', color='blue')

#negative gradients
plt.arrow(-6, 20, 0.3, -40, head_width=0.15, head_length=4, color='black')
plt.arrow(-2.5, 0, -1, -35, head_width=0.15, head_length=4, color='black')
plt.arrow(-1, 15, 0.85, -15, head_width=0.15, head_length=4, color='black')
plt.arrow(2, 40, -0.4, -30, head_width=0.15, head_length=4, color='black')

plt.xlabel('x')
plt.ylabel('y')
plt.xlim(-6.5, 2.5)
plt.ylim(-60, 70)
plt.legend()
plt.grid()
plt.savefig(os.path.join(dirpath, 'negative_gradients'))
plt.show()

File: slides/04-multivariate-first-order/rsrc/negative_gradients.R
# ------------------------------------------------------------------------------
# multivariate first order

# FIG: plot negative gradients
# ------------------------------------------------------------------------------

library(ggplot2)

# ------------------------------------------------------------------------------

# Define the function
f <- function(x) { x^4 + 7*x^3 + 5*x^2 - 17*x + 3 }

# Generate x and y values
x_vals <- seq(-6, 2, length.out = 50)
y_vals <- f(x_vals)

# Global minimum (from Python)
nsx <- -4.48
nsy <- f(nsx)

# Define vertical dashed lines
point1 <- data.frame(x = c(nsx, nsx), y = c(nsy, -60))
point2 <- data.frame(x = c(nsx, -6.5), y = c(nsy, nsy))

# Define negative gradient arrows
arrows <- data.frame(
  xstart = c(-6, -2.5, -1, 2),
  ystart = c(20, 0, 15, 40),
  xend = c(-6 + 0.3, -2.5 - 1, -1 + 0.85, 2 - 0.4),
  yend = c(20 - 40, 0 - 35, 15 - 15, 40 - 30)
)

# Create the plot
p <- ggplot() +
  # Plot function curve
  geom_line(data = data.frame(x = x_vals, y = y_vals), aes(x, y), color = "red") +
  
  # Add vertical dashed lines
  geom_line(data = point1, aes(x, y), linetype = "dashed", color = "blue") +
  geom_line(data = point2, aes(x, y), linetype = "dashed", color = "blue") +
  
  # Add negative gradient arrows
  geom_segment(data = arrows, aes(x = xstart, y = ystart, xend = xend, yend = yend), 
               arrow = arrow(length = unit(0.2, "inches")), color = "black") +
  annotate("text", x = 1, y = -50, label = expression(f(x) == x^4 + 7*x^3 + 5*x^2 - 17*x + 3), 
           hjust = 1, size = 5, color = "black") +
  labs(x = "x", y = "y") +
  
  xlim(-6.5, 2.5) + ylim(-60, 70) +  # Match x and y limits
  theme_minimal()

p
ggsave("../figure_man/negative_gradients.png", p, width = 6, height = 4)
